tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/VoteAPIHelper.java

```diff
@@ -0,0 +1,342 @@
+//                    r.section = section;
+                    r.localeDisplayName = locale.getDisplayName();
+                    r.dataLoadTime = et.toString();
+                    // ????
+//                    if (ctx.hasField("dashboard")) {
+//                        JSONArray issues = VettingViewerQueue.getInstance().getErrorOnPath(ctx.getLocale(), ctx, ctx.session, baseXp);
+//                        r.key("issues").value(issues);
+//                    }
+                    r.rows = calculateRows(section.getAll());
+                } catch (Throwable t) {
+                    SurveyLog.logException(t, "RefreshRow write");
+                    return new STError(t).build();
+                }
+
+            }
+          return Response.ok(r).build();
+        } catch (Throwable t) {
+            SurveyLog.logException(t, "Trying to load " + loc + " / " + xpath);
+            return new STError(t).build(); // 500
+        }
+    }
+
+    private static RowResponse.Row[] calculateRows(Collection<DataRow> all) {
+        List<RowResponse.Row> list = new LinkedList<>();
+        for(final DataRow r : all) {
+            RowResponse.Row row = new RowResponse.Row();
+            // from DataSection.DataRow.toJSONString()
+
+            // TODO: ourVote
+            row.xpath = XPathTable.getStringIDString(r.getXpath());
+
+            // Add candidate items
+            List<RowResponse.Row.Candidate> items = new LinkedList<>();
+            for(final CandidateItem i : r.items.values()) {
+                RowResponse.Row.Candidate c = new RowResponse.Row.Candidate();
+                c.value = i.getValue();
+                c.displayValue = i.getProcessedValue();
+                c.example = i.getExample();
+                c.isBaselineValue = i.isBaselineValue();
+
+                // TODO: pass underlying values to FE. Don't pass CSS class.
+                //    c.pClass = i.getPClass();
+                // TODO: c.tests = i.tests
+                items.add(c);
+            }
+            row.items = items.toArray(new RowResponse.Row.Candidate[items.size()]);
+            list.add(row);
+        }
+        return list.toArray(new RowResponse.Row[list.size()]);
+    }
+
+    static Response handleVote(String loc, String xpath, VoteRequest request, final CookieSession mySession) {
+        VoteResponse r = new VoteResponse();
+
+        mySession.userDidAction();
+
+        CLDRLocale locale = CLDRLocale.getInstance(loc);
+        CheckCLDR.Options options = DataSection.getOptions(null, mySession, locale); // TODO: why null for WebContext here?
+        final SurveyMain sm = CookieSession.sm;
+        STFactory stf = sm.getSTFactory();
+        TestResultBundle cc = stf.getTestResult(locale, options);
+        int id = sm.xpt.getXpathIdFromStringId(xpath);
+        String xp = sm.xpt.getById(id);
+        final List<CheckStatus> result = new ArrayList<>();
+        SurveyMain.UserLocaleStuff uf = null;
+        boolean dataEmpty = false;
```

Comment:
so many local variables, maybe they could go in a new class, which might facilitate dividing this function into shorter functions

Reply:
maybe, it still needs more 'pruning'…